src/core/model/Monitor.h

```diff
@@ -0,0 +1,44 @@
+#include <cassert>
+#include <mutex>
+
+template<class T>
+class Monitor
+{
+public:
+    template<typename ...Args>
+    Monitor(Args&&... args);
+
+    struct LockedMonitor
+    {
+        LockedMonitor(Monitor* monitor);
+        T* operator->();
+        void ReplaceModel(T model);
+        private:
+            Monitor* mon;
+            std::unique_lock<std::mutex> lock;
+    };
+
+    struct TryLockedMonitor
+    {
+        TryLockedMonitor(Monitor* monitor);
+        T* operator->();
+        void ReplaceModel(T model);
+        bool lockAcquired;
+        private:
+            Monitor* mon = nullptr;
+            std::unique_lock<std::mutex> lock;
+    };
+
+
+    LockedMonitor operator->();
+
+    LockedMonitor lock();
+
+    TryLockedMonitor tryLock();
```

Comment:
I think an optional will force the accessory to check / explicitly dereference the optional.
Also, `try_lock` should be written in snake case for convenience (c++ STL does so). 

```suggestion
    std::optional<LockedMonitor> try_lock();
```